common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -139,6 +140,20 @@ def unnest(self):
         """
         self._active_count -= 1
 
+    def add_publish_item(self, location):
+        """
+        Adds an item to the set of things that have been published
+        """
+        self._publish_items.add(location.version_agnostic().for_branch(None))
```

Comment:
@ormsbee 
So, I am keeping a list of items whose published state has changed - this allows for a more granular handling of the signal. Of course, once past a certain threshold, it is better to simply re-do the whole course index. I have currently left this up to the receiver to decide when they'd like to get less granular.

As you pointed out in the original PR, it is possible that this list of items could grow to be almost every block in the course. I believe that this is unlikely because the general approach is to mark only the topmost item, but perhaps course-imports would create a long list.

I am interested in getting your opinion on whether I should be adding some sort of gate here - e.g. if the list of items to publish get larger than 10 (arbitrary number) then perhaps this set simply steps back to only include the course key - this would avoid when the item list gets very large. In that case, the receiver would see the course key in the item list, and know that a full-course operation needs to be done (in my case a full course reindex) instead of just a partial.

A tricky wrench in the works here is upon deleting an item - either we add all of it's children to the list of changes, or have a check for "orphaned" index entries after each index operation (or some combination of the 2 based upon the upper threshold for the item list length).

Would love your feedback on this... thanks, Marty